examples/samd/usbserial/main.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2020, Erik Henriksson
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <tusb.h>
+
+#include <modm/board.hpp>
+#include <modm/io.hpp>
+#include <modm/processing.hpp>
+
+using namespace Board;
+
+static volatile uint32_t *DBL_TAP_PTR  (volatile uint32_t *)(HMCRAMC0_ADDR + HMCRAMC0_SIZE - 4);
+constexpr uint32_t DBL_TAP_MAGIC = 0xf01669ef;  // Defined in UF2 bootloader
+
+// Arduino style hack that resets to DFU on disconnect at 1200bps
+modm_extern_c void tud_cdc_line_state_cb(uint8_t itf, bool dtr, bool rts)
+{
+	if (!dtr)  // disconnected
+	{
+		cdc_line_coding_t coding;
+		tud_cdc_get_line_coding(&coding);
+
+		if (coding.bit_rate == 1200)
+		{
+			*DBL_TAP_PTR = DBL_TAP_MAGIC;
+			NVIC_SystemReset();
+		}
+	}
+}
+
+int main()
+{
+	Board::initialize();
+
+	// Output 500hz PWM on D12 so we can validate the GCLK0 clock speed.
+	PM->APBCMASK.reg |= PM_APBCMASK_TCC0;
+	TCC0->PER.bit.PER = SystemClock::Frequency / 1000;
+	TCC0->CC[3].bit.CC = SystemClock::Frequency / 2000;
+	TCC0->CTRLA.bit.ENABLE = true;
+	D12::Wo3<Peripheral::Tcc0>::connect();
+	GenericClockController::connect<ClockPeripheral::Tcc0>(
+	ClockGenerator::System);
+
+	Led::set();
+	Usb::initialize<Board::SystemClock>();
+	Usb::connect<GpioA24::Dm, GpioA25::Dp>();
```

Comment:
A nit on the API here, why does the user needs to know about `Dm` and `Dp` when initializing a USB peripheral?

I previously made this `Usb::connect<GpioA24, GpioA25>()` which has the same static safety but does not require the reader to fully understand a 1000+ pages datasheet.

I have a similar comment about Uart, where the user currently is required to know which pins map to which sercom pads, which at least I have no chance doing without digging datasheet or looking at the generated code (I just wanted to print hello world!). We could similarly avoid that by supporting something like `Uart0::connect<GpioA24, GpioA25>()`.

Reply:
Been there, done that. I have some very in-depth knowledge (\*sigh\*) about a number of issues with the signals API in general that needed to be solved:

1. You need a clear mapping of *at least* (Peripheral, Pin, Signal) to determine the correct bits to flip. In your case the Signal is given implicitly by argument position.

2. For the STM32F1 series, you can only remap groups of signals to overlapping groups of pins. Thus depending on the specific (Pin, Signal) group mapping at the callsite you must choose from multiple remap groups or static assert. Example from a generated file (`examples/blue_pill/blink/modm/src/modm/platform/gpio/connector_specialized.hpp:270`):

```cpp
static_assert(/* check if group mapping exists */,
    "This pin set contains conflicting remap groups!\nAvailable groups for Tim2 are:\n"
    " Tim2 |    0    |    1    |    2    |    3    \n"
    "------|---------|---------|---------|---------\n"
    "  A0  | Ch1,Etr |         | Ch1,Etr |         \n"
    "  A1  |   Ch2   |         |   Ch2   |         \n"
    "  A2  |   Ch3   |   Ch3   |         |         \n"
    "  A3  |   Ch4   |   Ch4   |         |         \n"
    " A15  |         | Ch1,Etr |         | Ch1,Etr \n"
    "  B3  |         |   Ch2   |         |   Ch2   \n"
    " B10  |         |         |   Ch3   |   Ch3   \n"
    " B11  |         |         |   Ch4   |   Ch4   \n"
);
```

3. You need a variable number of connect pins with some default behavior. For example, you can use SPI to bit-bang WS2812 LEDs, but you only need MOSI for that. Or perhaps you only need to read something with only MISO and SCK. This makes it difficult to use fixed Signal argument positions with defaults: 

```cpp
template<class Sck = GpioUnused, class Mosi = GpioUnused, class Miso = GpioUnused >
Spi::connect() {}

Spi::connect<GpioUnused, GpioA1>(); // For Mosi only
Spi::connect<GpioA0, GpioUnused, GpioA2>(); // for Sck, Miso only
```

4. Due to lack of named arguments, you don't actually know what pin is connected to what signal, particularly with defaulted arguments. This is much clearer:

```cpp
Spi::connect<GpioA1::Mosi>();
Spi::connect<GpioA0::Sck, GpioA2::Miso>();
```

5. It's unclear what and how many Signals exactly you need to connect for your use-case For example the FSMC can be configured at runtime in different connection modes, making it quite difficult to do this *without* variadic templates. This is still type-safe, but not nothing was explicitly coded for this config:
```cpp
GpioConnector<Peripheral::Fsmc,
    Board::display::fsmc::D0::D0,
    Board::display::fsmc::D1::D1,
    Board::display::fsmc::D2::D2,
    Board::display::fsmc::D3::D3,
    Board::display::fsmc::D4::D4,
    Board::display::fsmc::D5::D5,
    Board::display::fsmc::D6::D6,
    Board::display::fsmc::D7::D7,
    Board::display::fsmc::D8::D8,
    Board::display::fsmc::D9::D9,
    Board::display::fsmc::D10::D10,
    Board::display::fsmc::D11::D11,
    Board::display::fsmc::D12::D12,
    Board::display::fsmc::D13::D13,
    Board::display::fsmc::D14::D14,
    Board::display::fsmc::D15::D15,
    Board::display::fsmc::Noe::Noe,
    Board::display::fsmc::Nwe::Nwe,
    Board::display::fsmc::A18::A18
>::connect();
```

6. Finally, you can define pin configuration defaults easily with our current API. For example, you can use UART with Tx or Rx or Tx+Rx, but this code works for all three cases, since GpioUnused is returned from `GetSignal<Tx>` if the signal does not exist:

```cpp
template< template<Peripheral _> class... Signals >
static void
connect(Gpio::InputType InputTypeRx = Gpio::InputType::PullUp,
        Gpio::OutputType OutputTypeTx = Gpio::OutputType::PushPull)
{
    using Connector = GpioConnector<Peripheral::{{ name }}, Signals...>;
    using Tx = typename Connector::template GetSignal< Gpio::Signal::Tx >; // Find Tx signal in group
    using Rx = typename Connector::template GetSignal< Gpio::Signal::Rx >; // Find Rx signal in group
    static_assert(((Connector::template IsValid<Tx> and Connector::template IsValid<Rx>) and sizeof...(Signals) == 2) or
                  ((Connector::template IsValid<Tx> or  Connector::template IsValid<Rx>) and sizeof...(Signals) == 1),
                  "{{ name }}::connect() requires one Tx and/or one Rx signal!");

    Tx::setOutput(OutputTypeTx); // Does nothing if GpioUnused
    Rx::setInput(InputTypeRx); // Does nothing if GpioUnused
    Connector::connect();
}
```

What would be useful is a slim HTML-only version of a pinout picker, where you use modm-devices data to configure the signal<->pin map visually and then generate the appropriate modm code. That would be much easier to prototype complex dependencies like SystemClock configurations.

There's already a curses-based CLI that does some of this: https://github.com/tgree/stm_layout

Reply:
And another point is that the the STM32 devices do not have a crossbar connect, thus you cannot connect every pin to any signal. So you have no choice but to look into the datasheet, or use the graphical CubeMX to configure your signals.

Reply:
Thanks so much for the thorough reply!

I opened #481 for the followup discussion on this, I still think this can be improved in the light of SAM (where peripheral / signal / function is not enough to determine the correct wiring).